semgrep/semgrep/core_runner.py

```diff
@@ -99,6 +101,140 @@ def uniq_error_id(
     return (error.code, error.path, error.start, error.end, error.message)
 
 
+class StreamingSemgrepCore:
+    """
+    Handles running semgrep-core in a streaming fashion
+
+    This behavior is assumed to be that semgrep-core:
+    - prints a "." on a newline for every file it finishes scanning
+    - prints a single json blob of all results
+
+    Exposes the subprocess.CompletedProcess properties for
+    expediency in integrating
+    """
+
+    def __init__(self, cmd: List[str], total: int) -> None:
+        """
+        cmd: semgrep-core command to run
+        total: how many rules to run / how many "." we expect to see
+               used to display progress_bar
+        """
+        self._cmd = cmd
+        self._total = total
+        self._stdout = ""
+        self._stderr = ""
+        self._progress_bar: Optional[tqdm] = None  # type: ignore
+
+    @property
+    def stdout(self) -> str:
+        # stdout of semgrep-core sans "."
+        return self._stdout
+
+    @property
+    def stderr(self) -> str:
+        # stderr of semgrep-core command
+        return self._stderr
+
+    async def _core_stdout_processor(
+        self, stream: Optional[asyncio.StreamReader]
+    ) -> None:
+        """
+        Asynchronously process stdout of semgrep-core
+
+        Updates progress bar one increment for every "." it sees from semgrep-core
+        stdout
+
+        When it sees non-"." output it saves it to self._stdout
+        """
+        # appease mypy. stream is only None if call to create_subproccess_exec
+        # sets stdout/stderr stream to None
+        assert stream
+        while True:
+            # blocking read if buffer doesnt contain any lines or EOF
+            line_bytes = await stream.readline()
+
+            # readline returns None on EOF
+            if not line_bytes:
+                break
+
+            line = line_bytes.decode("utf-8")
+            if line.strip() == ".":
+                if self._progress_bar:
+                    self._progress_bar.update()
+            else:
+                self._stdout += line
+
+    async def _core_stderr_processor(
+        self, stream: Optional[asyncio.StreamReader]
+    ) -> None:
+        """
+        Asynchronously process stderr of semgrep-core
+
+        Basically works sychronnously and combines output to
+        stderr to self._stderr
+        """
+        assert stream
+        while True:
+            # blocking read if buffer doesnt contain any lines or EOF
+            line_bytes = await stream.readline()
+
+            # readline returns None on EOF
+            if not line_bytes:
+                break
+
+            line = line_bytes.decode("utf-8")
+            self._stderr += line
+
+    async def _stream_subprocess(self) -> int:
+        process = await asyncio.create_subprocess_exec(
+            *self._cmd,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            limit=1024 * 1024 * 1024,
```

Comment:
[The default is 64KiB](https://docs.python.org/3/library/asyncio-stream.html#:~:text=The%20returned%20reader%20and%20writer,is%20set%20to%2064%20KiB.).

Are we should 1 GiB is safe here? it's way higher than I'd expect, and based on the naming, I don't think this is the limit for maximum transferred data anyway

Reply:
the default was way too small for some cases of semgrep-core output since I think asyncio here blocks until a newline but the json output could be pretty large